src/runtime/engines/singularity/container.go

```diff
@@ -1001,27 +1032,42 @@ func (c *container) addUserbindsMount(system *mount.System) error {
 			}
 		}
 
-		sylog.Debugf("Adding %s to mount list\n", src)
-		if src == "/dev" {
-			// special case for /dev mount to override default mount behaviour
-			// by example with --contain option
-			if c.engine.EngineConfig.File.MountDev == "yes" {
-				system.Points.RemoveByTag(mount.DevTag)
-
-				err := system.Points.AddBind(mount.DevTag, "/dev", dst, syscall.MS_BIND|syscall.MS_REC)
-				if err != nil {
-					return fmt.Errorf("unable to add dev to mount list: %s", err)
+		// special case for /dev mount to override default mount behaviour
+		// with --contain option or 'mount dev = minimal'
+		if strings.HasPrefix(src, "/dev") {
+			if c.engine.EngineConfig.File.MountDev == "minimal" || c.engine.EngineConfig.GetContain() {
+				if strings.HasPrefix(src, "/dev/shm") || strings.HasPrefix(src, "/dev/mqueue") {
+					sylog.Warningf("Skipping %s bind mount: not authorized", src)
+				}
```

Comment:
A fresh /dev/shm is already mounted so a request to bind that should get rejected as a duplicate bind, right?

As for /dev/mqueue, I don't see why someone would want to do that if they want to restrict the container, but if they ask for it, why should singularity prevent it?

Reply:
The main reason is that these mount points are world writable and don't want to deal with nasty symlinks for security reasons, so Singularity just prevent bind and display a `not authorized` message